### PR TITLE
Model Float on Hoistables semantics

### PR DESCRIPTION
**Pull request by gnoff**
## Hoistables

In the original implementation of Float, all hoisted elements were treated like Resources. They had deduplication semantics and hydrated based on a key. This made certain kinds of hoists very challenging such as sequences of meta tags for `og:image:...` metadata. The reason is each tag along is not dedupable based on only it's intrinsic properties. two identical tags may need to be included and hoisted together with preceding meta tags that describe a semantic object with a linear set of html nodes.

It was clear that the concept of Browser Resources (stylesheets / scripts / preloads) did not extend universally to all hositable tags (title, meta, other links, etc...)

Additionally while Resources benefit from deduping they suffer an inability to update because while we may have multiple rendered elements that refer to a single Resource it isn't unambiguous which element owns the props on the underlying resource. We could try merging props, but that is still really hard to reason about for authors. Instead we restrict Resource semantics to freezing the props at the time the Resource is first constructed and warn if you attempt to render the same Resource with different props via another rendered element or by updating an existing element for that Resource.

This lack of updating restriction is however way more extreme than necessary for instances that get hoisted but otherwise do not dedupe; where there is a well defined DOM instance for each rendered element. We should be able to update props on these instances.

Hoistable is a generalization of what Float tries to model for hoisting. Instead of assuming every hoistable element is a Resource we now have two distinct categories, hoistable elements and hoistable resources. As one might guess the former has semantics that match regular Host Components except the placement of the node is usually in the <head>. The latter continues to behave how the original implementation of HostResource behaved with the first iteration of Float

### Hoistable Element
On the server hoistable elements render just like regular tags except the output is stored in special queues that can be emitted in the stream earlier than they otherwise would be if rendered in place. This also allow for instance the ability to render a hoistable before even rendering the <html> tag because the queues for hoistable elements won't flush until after we have flushed the preamble (`<DOCTYPE html><html><head>`). 

On the client, hoistable elements largely operate like HostComponents. The most notable difference is in the hydration strategy. If we are hydrating and encounter a hoistable element we will look for all tags in the document that could potentially be a match and we check whether the attributes match the props for this particular instance. We also do this in the commit phase rather than the render phase. The reason  hydration can be done for HostComponents in render is the instance will be removed from the document if hydration fails so mutating it in render is safe. For hoistables the nodes are not in a hydration boundary (Root or SuspenseBoundary at time of writing) and thus if hydration fails and we may have an instance marked as bound to some Fiber when that Fiber never commits. Moving the hydration matching to commit ensures we will always succeed in pairing the hoisted DOM instance with a Fiber that has committed.

### Hoistable Resource
On the server and client the semantics of Resources are largely the same they just don't apply to title, meta, and most link tags anymore. Resources hoist and dedupe via an `href` key and are ref counted. In a future update we will add a garbage collector so we can clean up Resources that no longer have any references

## `<style>` support
In earlier implementations there was no support for <style> tags. This PR adds support for treating `<style href="..." precedence="...">...</style>` as a Resource analagous to `<link rel="stylesheet" href="..." precedence="..." />`

It may seem odd at first to require an href to get Resource semantics for a style tag. The rationale is that these are for inlining of actual external stylesheets as an optimization and for URI like scoping of inline styles for css-in-js libraries. The href indicates that the key space for `<style>` and `<link rel="stylesheet" />` Resources is shared. and the precedence is there to allow for interleaving of both kinds of Style resources. This is an advanced feature that we do not expect most app developers to use directly but will be quite handy for various styling libraries and for folks who want to inline as much as possible once Fizz supports this feature.

## refactor notes
* HostResource Fiber type is renamed HostHoistable to reflect the generalization of the concept
* The Resource object representation is modified to reduce hidden class checks and to use less memory overall
* The thing that distinguishes a resource from an element is whether the Fiber has a memoizedState. If it does, it will use resource semantics, otherwise element semantics
* The time complexity of matching hositable elements for hydration should be improved

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -1624,19 +1629,33 @@ function updateHostComponent(
...
+  const resource = (workInProgress.memoizedState = getResource(
     workInProgress.type,
-    workInProgress.pendingProps,
     currentProps,
-  );
+    workInProgress.pendingProps,
+  ));
+  if (current === null) {
+    if (!getIsHydrating() && resource === null) {
+      // This is not a Resource Hoistable and we aren't hydrating so we construct the
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberBeginWork.js:1646]: construct the... what?

File: packages/react-dom/src/__tests__/ReactDOMComponent-test.js [hunk 2]
```diff
@@ -494,6 +494,7 @@ describe('ReactDOMComponent', () => {
             'To fix this, either do not render the element at all ' +
             'or pass null to href instead of an empty string.',
         );
+        console.log('container', container.outerHTML);
```
Comment (sebmarkbage) [inline packages/react-dom/src/__tests__/ReactDOMComponent-test.js:497]: rm

File: packages/react-dom-bindings/src/client/ReactDOMFloatClient.js [hunk 3]
```diff
@@ -1390,40 +802,220 @@ function insertStyleInstance(
...
+// --------------------------------------
+
+export function hydrateHoistable(
+  hoistableRoot: HoistableRoot,
+  type: HoistableTagType,
+  props: any,
+  internalInstanceHandle: Object,
+): Instance {
+  const ownerDocument = getDocumentFromRoot(hoistableRoot);
+  const nodes = ownerDocument.querySelectorAll(type);
```
Comment (sebmarkbage) [inline packages/react-dom-bindings/src/client/ReactDOMFloatClient.js:846]: `getElementsByTagName` by-passes the parsing.

File: packages/react-dom-bindings/src/client/ReactDOMFloatClient.js [hunk 4]
```diff
@@ -1390,40 +802,220 @@ function insertStyleInstance(
...
+    }
+    childString = '' + (child: any);
   } else {
-    throw new Error(
-      'While attempting to insert a Resource, React expected the Document to contain' +
-        ' a head element but it was not found.',
-    );
+    childString = '';
+  }
+  nodeLoop: for (let i = 0; i < nodes.length; i++) {
```
Comment (sebmarkbage) [inline packages/react-dom-bindings/src/client/ReactDOMFloatClient.js:868]: I'm not loving the computational complexity of N * N queries for number of nodes. I wonder if we need to do something like caching in a Map.

**Review by sebmarkbage — APPROVED** [review]
This feels like the right direction. We can nit pick later but let's get this in sooner rather than later because it'll be a pain to rebase later for everyone involved.

Fix the lints and add a more comprehensive description to others and our future self what's going here, before landing.

File: packages/react-dom-bindings/src/shared/ReactDOMResourceValidation.js [hunk 5]
```diff
@@ -755,24 +179,429 @@ export function validatePreinitArguments(href: mixed, options: mixed) {
...
+    [string]: any,
+  },
+  different: {
+    [string]: {
+      original: any,
+      latest: any,
+    },
+  },
+};
+export function compareResourcePropsForWarning(
```
[the remark this answers is not in the document]
Comment (gnoff) [reply packages/react-dom-bindings/src/shared/ReactDOMResourceValidation.js:220]: yeah I already had to do some selective processing of the diff for some cases like if you don't pass a precedence in preinit that still is compatible with a "default" precedence on a stylesheet link. I'll see about refactoring it in a future update to be more direct

File: packages/react-dom-bindings/src/client/ReactDOMFloatClient.js (hunk 4 above)
Comment (gnoff) [reply packages/react-dom-bindings/src/client/ReactDOMFloatClient.js:868]: yeah same, i'll think about more efficient implementations for a follow up

File: packages/react-dom-bindings/src/client/ReactDOMFloatClient.js [hunk 6]
```diff
@@ -245,10 +144,15 @@ function getDocumentForPreloads(): ?Document {
...
 }
 
-function getDocumentFromRoot(root: FloatRoot): Document {
+function getDocumentFromRoot(root: HoistableRoot): Document {
   return root.ownerDocument || root;
 }
 
+export function getHoistableRoot(container: Container): HoistableRoot {
+  // Flow thinks getRootNode returns Node but we know it is actualy either a Document or ShadowRoot
+  return ((container.getRootNode(): any): Document | ShadowRoot);
```
Comment (poteto) [inline packages/react-dom-bindings/src/client/ReactDOMFloatClient.js:153]: @gnoff should we be using the `getRootNode` helper from above here? This is throwing a bunch of errors in internal classic component tests

**Comment by rkurbatov** [comment]
Wow. Some magic as usual :-P

**Pull request merged**